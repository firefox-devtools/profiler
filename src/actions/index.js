/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileView from './profile-view';
import * as publish from './publish';
import * as receiveProfile from './receive-profile';
import * as zippedProfiles from './zipped-profiles';
import * as code from './code';

export default Object.assign(
  {},
  app,
  icons,
  profileView,
  publish,
  receiveProfile,
  zippedProfiles,
  code
);
