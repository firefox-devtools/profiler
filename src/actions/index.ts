/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
import * as app from './app';
import * as icons from './icons';
import * as l10n from './l10n';
import * as profileView from './profile-view';
import * as publish from './publish';
import * as receiveProfile from './receive-profile';
import * as zippedProfiles from './zipped-profiles';
import * as code from './code';

export default Object.assign(
  {},
  app,
  icons,
  l10n,
  profileView,
  publish,
  receiveProfile,
  zippedProfiles,
  code
);
