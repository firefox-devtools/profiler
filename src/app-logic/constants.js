/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import type { MarkerPhase } from 'firefox-profiler/types';

// The current version of the Gecko profile format.
// Please don't forget to update the gecko profile format changelog in
// `docs-developer/CHANGELOG-formats.md`.
export const GECKO_PROFILE_VERSION = 31;

// The current version of the "processed" profile format.
// Please don't forget to update the processed profile format changelog in
// `docs-developer/CHANGELOG-formats.md`.
export const PROCESSED_PROFILE_VERSION = 55;

// The following are the margin sizes for the left and right of the timeline. Independent
// components need to share these values.
export const TIMELINE_MARGIN_RIGHT = 15;
export const TIMELINE_MARGIN_LEFT = 150;
export const ACTIVE_TAB_TIMELINE_MARGIN_LEFT = 0;

// Export the value for tests, and for computing the max height of the timeline
// for the splitter.
export const FULL_TRACK_SCREENSHOT_HEIGHT = 50;
export const ACTIVE_TAB_TRACK_SCREENSHOT_HEIGHT = 30;

// The following values are for network track.
export const TRACK_NETWORK_ROW_HEIGHT = 5;
export const TRACK_NETWORK_ROW_REPEAT = 7;
export const TRACK_NETWORK_HEIGHT =
  TRACK_NETWORK_ROW_HEIGHT * TRACK_NETWORK_ROW_REPEAT;

// The following values are for memory track.
export const TRACK_MEMORY_GRAPH_HEIGHT = 25;
export const TRACK_MEMORY_MARKERS_HEIGHT = 15;
export const TRACK_MEMORY_HEIGHT =
  TRACK_MEMORY_GRAPH_HEIGHT + TRACK_MEMORY_MARKERS_HEIGHT;
export const TRACK_MEMORY_LINE_WIDTH = 2;
export const TRACK_MEMORY_DEFAULT_COLOR = 'orange';

// The following values are for the bandwidth track.
export const TRACK_BANDWIDTH_HEIGHT = 25;
export const TRACK_BANDWIDTH_LINE_WIDTH = 2;
export const TRACK_BANDWIDTH_DEFAULT_COLOR = 'blue';

// The following values are for experimental event delay track.
export const TRACK_EVENT_DELAY_HEIGHT = 40;
export const TRACK_EVENT_DELAY_LINE_WIDTH = 2;

// The following values are for IPC track.
export const TRACK_IPC_MARKERS_HEIGHT = 25;
export const TRACK_IPC_HEIGHT = TRACK_IPC_MARKERS_HEIGHT;

// The following values are the defaults for marker tracks
export const TRACK_MARKER_HEIGHT = 25;
export const TRACK_MARKER_LINE_WIDTH = 2;
export const TRACK_MARKER_DEFAULT_COLOR = 'grey';

// Height of the blank area in process track.
export const TRACK_PROCESS_BLANK_HEIGHT = 30;

// Height of timeline ruler.
export const TIMELINE_RULER_HEIGHT = 20;

// Height of the power track.
export const TRACK_POWER_HEIGHT = 25;
export const TRACK_POWER_LINE_WIDTH = 2;
export const TRACK_POWER_DEFAULT_COLOR = 'grey';

// Height of the process cpu track.
export const TRACK_PROCESS_CPU_HEIGHT = 25;
export const TRACK_PROCESS_CPU_LINE_WIDTH = 2;

// JS Tracer has very high fidelity information, and needs a more fine-grained zoom.
export const JS_TRACER_MAXIMUM_CHART_ZOOM = 0.001;

// The following values are for the visual progress tracks.
export const TRACK_VISUAL_PROGRESS_HEIGHT = 40;
export const TRACK_VISUAL_PROGRESS_LINE_WIDTH = 2;

// Height of the active tab resources panel header.
export const ACTIVE_TAB_TIMELINE_RESOURCES_HEADER_HEIGHT = 20;

// =============================================================================
// Storage and server-related constants
// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// For the 2 values GOOGLE_STORAGE_BUCKET and PROFILER_SERVER_ORIGIN, several
// values are possible, so that you can easily switch between existing server
// (both local or remote).
//
// GOOGLE_STORAGE_BUCKET
// ---------------------
// This defines which bucket we fetch profile data at load time.

// Google storage bucket, where production profile data is stored:
export const GOOGLE_STORAGE_BUCKET = 'profile-store';

// You can also use one of the following values instead:
// To use the bucket used by the server deployment for the main branch:
// export const GOOGLE_STORAGE_BUCKET = 'moz-fx-dev-firefoxprofiler-bucket';

// To use the bucket developers usually use on their local working copy:
// export const GOOGLE_STORAGE_BUCKET = 'profile-store-julien-dev';

// PROFILER_SERVER_ORIGIN
// ----------------------
// This defines our server-side endpoint. This is currently used to publish
// profiles and manage shortlinks.

// This is the production server:
export const PROFILER_SERVER_ORIGIN = 'https://api.profiler.firefox.com';

// This is the deployment from the main branch:
// export const PROFILER_SERVER_ORIGIN = 'https://dev.firefoxprofiler.nonprod.cloudops.mozgcp.net';

// This is your local server:
// export const PROFILER_SERVER_ORIGIN = 'http://localhost:5252';

// SYMBOL_SERVER_URL
// -----------------
// Can be overridden with the URL parameter `symbolServer=SERVERURL`.
// You can change this to run a local symbol server (for example using profiler-symbol-server [1])
// and set it to e.g. 'http://localhost:8000/'.
//
// [1] https://github.com/mstange/profiler-symbol-server/

// This is the default server.
export const SYMBOL_SERVER_URL = 'https://symbolication.services.mozilla.com';

// See the MarkerPhase type for more information.
export const INSTANT: MarkerPhase = 0;
export const INTERVAL: MarkerPhase = 1;
export const INTERVAL_START: MarkerPhase = 2;
export const INTERVAL_END: MarkerPhase = 3;
