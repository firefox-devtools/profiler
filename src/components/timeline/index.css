/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.timelineOverflowEdgeIndicatorScrollbox {
  position: absolute;

  /* if the platform scrollbar ends up being larger than --vertical-scrollbar-reserved-width (which is unexpected), make sure there is no horizontal scrollbar */
  overflow: hidden auto;
  width: 100%;
  height: 100%;
  padding-left: var(--thread-label-column-width);
  margin-right: calc(var(--vertical-scrollbar-reserved-width) * -1);
  margin-left: calc(var(--thread-label-column-width) * -1);
  background: linear-gradient(
    to left,
    white,
    #eee calc(var(--vertical-scrollbar-reserved-width) - 1px),
    var(--grey-30) 0,
    var(--grey-30) var(--vertical-scrollbar-reserved-width),
    transparent 0
  );
}

.timelineThreadList {
  width: calc(100vw - var(--vertical-scrollbar-reserved-width));
  padding: 0;
  margin: 0 0 0 calc(var(--thread-label-column-width) * -1);
  background-color: var(--grey-20);
  box-shadow: inset 0 1px var(--grey-30);
  list-style: none;
}

.timelineSettingsHiddenTracks {
  /* Make sure and override all rules that have to do with button defaults. */
  position: relative;
  width: var(--thread-label-column-width);

  /* 8px is intentional to align this text with the text above. */
  padding: 0 20px 0 8px;
  border: 0;

  /* Separator line between tracks menu button and timeline ruler */
  border-right: 1px solid var(--grey-30);
  background: none;
  color: inherit;
  font: inherit;
  text-align: left;
}

.timelineSettingsHiddenTracks:hover {
  background-color: rgb(0 0 0 / 0.1);
}

.timelineSettingsHiddenTracks:hover:active {
  background-color: rgb(0 0 0 / 0.2);
}

/* This is the dropdown arrow on the left of the button. */
.timelineSettingsHiddenTracks::before {
  position: absolute;
  top: 2px;
  right: 2px;
  border-top: 6px solid;
  border-right: 4px solid transparent;
  border-bottom: 0 solid transparent;
  border-left: 4px solid transparent;
  margin-top: 5px;
  margin-right: 5px;
  margin-left: 5px;
  color: var(--grey-60);
  content: '';

  /* Opt-out of forced colors so arrow is visible in High Contrast Mode */
  forced-color-adjust: none;
}

.timelineSettingsHiddenTracksNumber {
  font-weight: bold;
}

.timelineHeader {
  /* This should be on top of the overflow edge indicator that wraps the thread
   list, so border bottom will always be visible. */
  z-index: 1;
  display: flex;
  height: 20px;
  flex-direction: row;

  /* Timeline header content should end before the scrollbar but also border
   * bottom should extend to the whole timeline header. Because of that, we are
   * creating a padding and a minus margin with the scrollbar width to make
   * this work. */
  padding-right: var(--vertical-scrollbar-reserved-width);
  border-bottom: 1px solid var(--grey-30);
  margin-right: calc(var(--vertical-scrollbar-reserved-width) * -1);
  margin-left: calc(var(--thread-label-column-width) * -1);

  /* This div needs to be apaque so that it can properly hide the hidden tracks
   * container border below */
  background: white;
}

.tracksContainer {
  /* This should be below the timeline header. */
  z-index: 0;

  /* Moving the timeline 1 pixel below the timeline header. This way, the first
   * visible track's top border will be hidden as expected. */
  margin-top: -1px;
  margin-bottom: -1px;
}

@media (forced-colors: active) {
  .timelineSettingsHiddenTracks {
    border: 1px solid currentcolor;
    background-color: ButtonFace;
    color: ButtonText;
  }

  .timelineSettingsHiddenTracks:hover {
    background-color: SelectedItemText;
    color: SelectedItem;
  }

  .timelineSettingsHiddenTracks:hover:active {
    border-color: ButtonText;
  }

  .timelineSettingsHiddenTracks::before {
    color: currentcolor;
  }
}
