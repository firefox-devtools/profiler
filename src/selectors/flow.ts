/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';

import {
  computeProfileFlowInfo,
  computeFlowTiming,
} from '../profile-logic/marker-data';
import { getThreadSelectors } from './per-thread';
import { getActiveFlows } from './url-state';
import type { ThreadSelectors } from './per-thread';
import {
  getThreads,
  getMarkerSchema,
  getRawProfileSharedData,
} from './profile';

import type {
  Selector,
  State,
  MarkerIndex,
  Marker,
  ProfileFlowInfo,
  FlowTiming,
} from 'firefox-profiler/types';

function _arraysShallowEqual(arr1: any[], arr2: any[]): boolean {
  return arr1.length === arr2.length && arr1.every((val, i) => val === arr2[i]);
}

function _createSelectorForAllThreads<T>(
  f: (selectors: ThreadSelectors, state: State) => T
): Selector<T[]> {
  let previousOutputPerThread: T[] = [];
  return function recomputeSelectorForAllThreads(state: State): T[] {
    const threads = getThreads(state);
    let outputPerThread = threads.map((_thread, i) => {
      const threadSelectors = getThreadSelectors(i);
      return f(threadSelectors, state);
    });
    if (_arraysShallowEqual(outputPerThread, previousOutputPerThread)) {
      outputPerThread = previousOutputPerThread;
    }
    previousOutputPerThread = outputPerThread;
    return outputPerThread;
  };
}

export const getFullMarkerListPerThread: Selector<Marker[][]> =
  _createSelectorForAllThreads(({ getFullMarkerList }, state) =>
    getFullMarkerList(state)
  );

export const getMarkerChartLabelGetterPerThread: Selector<
  Array<(marker: MarkerIndex) => string>
> = _createSelectorForAllThreads(({ getMarkerChartLabelGetter }, state) =>
  getMarkerChartLabelGetter(state)
);

export const getProfileFlowInfo: Selector<ProfileFlowInfo> = createSelector(
  getFullMarkerListPerThread,
  getThreads,
  getMarkerSchema,
  getRawProfileSharedData,
  computeProfileFlowInfo
);

export const getFlowTiming: Selector<FlowTiming> = createSelector(
  getProfileFlowInfo,
  getActiveFlows,
  computeFlowTiming
);
