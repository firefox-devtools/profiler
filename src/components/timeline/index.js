/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import { FullTimeline } from 'firefox-profiler/components/timeline/FullTimeline';

type TimelineProps = {||};

export class Timeline extends PureComponent<TimelineProps> {
  // This may contain a function that's called whenever we want to remove the
  // "wheel" listener.
  _removeWheelListener: null | (() => mixed) = null;

  // This effectively disable the pinch-to-zoom as well as ctrl+mousewheel
  // gestures. Indeed in the timeline it is confusing. In the future we'll want
  // to couple this with the preview selection like in the Viewport HOC.
  preventPinchToZoom(e: WheelEvent) {
    if (e.ctrlKey) {
      e.preventDefault();
    }
  }

  // This will be registered as a ref property to the DOM element displaying the
  // tracks. We use this solution of registering the wheel event in this ref
  // listener because:
  // * we can't use React's event handling, because it doesn't allow us to use
  //   the "passive: false" way of registering the event handler. But we need this
  //   if we want to be able to prevent the default action of page zooming.
  // * we want to be sure to register it whenever the element changes.
  _onTimelineMountWithRef = (ref: HTMLElement | null) => {
    if (this._removeWheelListener) {
      this._removeWheelListener();
      this._removeWheelListener = null;
    }

    if (!ref) {
      return;
    }

    // without pinning to a const variable, Flow isn't sure that we don't change
    // the `ref` variable in some of the function calls below, and therefore
    // that it won't be null.
    const existingRef = ref;

    // Disable pinch-to-zoom and ctrl + wheel otherwise, on the timeline.
    // Indeed the users are used to this gesture  to zoom in in our charts, and
    // may use the same gesture elsewhere because of their habits, however this
    // doesn't do what they expect and instead zooms in the page, which is distracting.
    existingRef.addEventListener('wheel', this.preventPinchToZoom, {
      passive: false,
    });

    this._removeWheelListener = () => {
      existingRef.removeEventListener('wheel', this.preventPinchToZoom, {
        passive: false,
      });
    };
  };

  componentWillUnmount() {
    if (this._removeWheelListener) {
      this._removeWheelListener();
      this._removeWheelListener = null;
    }
  }

  render() {
    return <FullTimeline innerElementRef={this._onTimelineMountWithRef} />;
  }
}
