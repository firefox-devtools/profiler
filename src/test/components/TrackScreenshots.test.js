/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type {
  Profile,
  RawThread,
  RawProfileSharedData,
  IndexIntoRawMarkerTable,
} from 'firefox-profiler/types';

import * as React from 'react';
import { Provider } from 'react-redux';

import {
  render,
  fireEvent,
  screen,
  act,
} from 'firefox-profiler/test/fixtures/testing-library';
import { commitRange } from '../../actions/profile-view';
import { TimelineTrackScreenshots } from '../../components/timeline/TrackScreenshots';
import { Timeline } from '../../components/timeline';
import { StringTable } from '../../utils/string-table';
import { ensureExists } from '../../utils/flow';
import { FULL_TRACK_SCREENSHOT_HEIGHT } from '../../app-logic/constants';

import { autoMockCanvasContext } from '../fixtures/mocks/canvas-context';
import { mockRaf } from '../fixtures/mocks/request-animation-frame';
import { storeWithProfile } from '../fixtures/stores';
import {
  getMouseEvent,
  addRootOverlayElement,
  removeRootOverlayElement,
  fireFullClick,
} from '../fixtures/utils';
import { getScreenshotTrackProfile } from '../fixtures/profiles/processed-profile';
import { getProfileWithNiceTracks } from '../fixtures/profiles/tracks';
import { getPreviewSelection } from '../../selectors/profile';
import { autoMockDomRect } from 'firefox-profiler/test/fixtures/mocks/domrect.js';
import {
  autoMockElementSize,
  setMockedElementSize,
} from '../fixtures/mocks/element-size';
import { autoMockIntersectionObserver } from '../fixtures/mocks/intersection-observer';

// Mock out the element size to have a 400 pixel width and some left/top
// positioning.
const TRACK_WIDTH = 400;
const LEFT = 100;
const TOP = 7;
const INITIAL_ELEMENT_SIZE = {
  width: TRACK_WIDTH,
  height: FULL_TRACK_SCREENSHOT_HEIGHT,
  offsetX: LEFT,
  offsetY: TOP,
};

describe('timeline/TrackScreenshots', function () {
  autoMockDomRect();
  autoMockCanvasContext();
  autoMockElementSize(INITIAL_ELEMENT_SIZE);
  autoMockIntersectionObserver();

  beforeEach(addRootOverlayElement);
  afterEach(removeRootOverlayElement);

  it('matches the component snapshot', () => {
    const { container, unmount } = setup();
    expect(container.firstChild).toMatchSnapshot();
    // Trigger any unmounting behavior handlers, just make sure it doesn't
    // throw any errors.
    unmount();
  });

  it('shows a hover when moving the mouse', () => {
    const { screenshotHover, moveMouseAndGetImageSize } = setup();

    expect(screenshotHover).toThrow();
    const size = moveMouseAndGetImageSize(LEFT + 0);
    expect(screenshotHover()).toBeTruthy();
    expect(size).toEqual({ width: 350, height: 175 });
  });

  it('sets a preview selection when clicking with the mouse', () => {
    const { selectionOverlay, screenshotClick, getState } = setup(
      undefined,
      <Timeline />
    );
    expect(selectionOverlay).toThrow();
    screenshotClick(LEFT);

    const expectedPreviewSelection = {
      hasSelection: true,
      isModifying: false,
      selectionEnd: 1,
      selectionStart: 0,
    };
    expect(getPreviewSelection(getState())).toEqual(expectedPreviewSelection);
    expect(selectionOverlay()).toBeTruthy();
  });

  it(`displays a smaller screenshot when preview selecting, and doesn't change it when clicking if a selection is already present`, () => {
    const {
      selectionOverlay,
      screenshotTrack,
      moveMouseAndGetImageSize,
      getState,
    } = setup(undefined, <Timeline />);

    const track = screenshotTrack();
    expect(selectionOverlay).toThrow();

    // Mousedown then Mousemove will do a selection.
    fireEvent(track, getMouseEvent('mousedown', { pageX: LEFT, pageY: TOP }));
    const screenShotSize = moveMouseAndGetImageSize(LEFT + 200);

    expect(selectionOverlay()).toBeTruthy();

    // And we should see a small screenshot hover while preview selecting.
    expect(screenShotSize).toEqual({ width: 100, height: 50 });

    // click should keep this existing selection
    const previouslySelectedPreviewSelection = getPreviewSelection(getState());
    fireEvent(
      track,
      getMouseEvent('mouseup', { pageX: LEFT + 200, pageY: TOP })
    );
    fireEvent(track, getMouseEvent('click', { pageX: LEFT + 200, pageY: TOP }));

    expect(getPreviewSelection(getState())).toEqual({
      ...previouslySelectedPreviewSelection,
      isModifying: false,
    });
    expect(selectionOverlay()).toBeTruthy();

    // but the screenshot hover size should now be bigger
    const screenshotSize = moveMouseAndGetImageSize(LEFT + 210);
    expect(screenshotSize).toEqual({ width: 350, height: 175 });
  });

  it('removes the hover when moving the mouse out', () => {
    const { screenshotHover, screenshotTrack, moveMouse } = setup();

    expect(screenshotHover).toThrow();

    moveMouse(LEFT + 0);
    expect(screenshotHover()).toBeTruthy();

    fireEvent.mouseLeave(screenshotTrack());
    expect(screenshotHover).toThrow();
  });

  it('moves the hover when moving the mouse', () => {
    const { moveMouseAndGetLeft } = setup();
    //Considering base to be 150, this value is big enough as a base
    const base = moveMouseAndGetLeft(150);
    // LEFT is 100 and the zoomed in screenshot's width is 350px.
    // So hovering between 100 and 175px should give a result of left == 0.
    // As soon as we pass 175px the left value should also move up.
    expect(moveMouseAndGetLeft(175 + 10)).toBe(base + 10);
    expect(moveMouseAndGetLeft(175 + 20)).toBe(base + 20);
  });

  it('places the hover image in the center of the track if there is enough space', () => {
    const { moveMouseAndGetTop } = setup();
    const containerTop = 100;
    const screenshotHeight = 175; // This is the height of the zoomed-in screenshot.
    setMockedElementSize({
      ...INITIAL_ELEMENT_SIZE,
      offsetX: LEFT, // repeating this property for more clarity
      offsetY: containerTop,
    });
    const pageX = LEFT;
    const expectedTop = Math.floor(
      containerTop + FULL_TRACK_SCREENSHOT_HEIGHT / 2 - screenshotHeight / 2
    );
    expect(moveMouseAndGetTop(pageX)).toBe(expectedTop);
  });

  it('makes sure the hover image does not go off the end of the container', () => {
    const { moveMouseAndGetLeft } = setup();
    const pageX = LEFT + TRACK_WIDTH - 1;
    expect(pageX > moveMouseAndGetLeft(pageX)).toBe(true);
  });

  it('makes sure the hover image does not go off the left side of screen', () => {
    const { moveMouseAndGetLeft } = setup();

    // Because the zoomed in hover screenshot's size is 350px, it's stuck at the
    // left of the window when the mouse is between 100 (the left of
    // the track) and 175px.
    expect(moveMouseAndGetLeft(100)).toBe(0);
    expect(moveMouseAndGetLeft(150)).toBe(0);
    expect(moveMouseAndGetLeft(175)).toBe(0);

    // Starting at 176px, the hover tooltip starts to move.
    expect(moveMouseAndGetLeft(176)).toBe(1);
  });

  it('makes sure the hover image does not go off the top side of screen', () => {
    const { moveMouseAndGetTop } = setup();
    const pageX = LEFT;
    expect(moveMouseAndGetTop(pageX)).toBe(0);
  });

  it('renders a screenshot images when zooming into a range without a screenshot start time actually in the range', () => {
    const profile = getScreenshotTrackProfile();
    const { shared, threads } = profile;
    const [thread] = threads;
    const markerIndexA = thread.markers.length - 3;
    const markerIndexB = thread.markers.length - 2;
    // We keep the last marker so that the profile's root range is correct.

    _setScreenshotMarkersToUnknown(thread, shared, markerIndexA, markerIndexB);

    const { dispatch, container } = setup(profile);
    act(() => {
      dispatch(
        commitRange(
          ensureExists(thread.markers.startTime[markerIndexA]),
          ensureExists(thread.markers.startTime[markerIndexB])
        )
      );
    });

    const firstImage = ensureExists(
      container.querySelector('.timelineTrackScreenshotImgContainer'),
      `Couldn't find at least one screenshot image.`
    );
    expect(parseInt(firstImage.style.left)).toBeGreaterThanOrEqual(0);
  });

  it('renders a no images when zooming into a range before screenshots', () => {
    const profile = getScreenshotTrackProfile();
    const { shared, threads } = profile;
    const [thread] = threads;

    const markerIndexA = 0;
    const markerIndexB = 1;

    _setScreenshotMarkersToUnknown(thread, shared, markerIndexA, markerIndexB);

    const { dispatch, container } = setup(profile);
    act(() => {
      dispatch(
        commitRange(
          ensureExists(thread.markers.startTime[markerIndexA]),
          ensureExists(thread.markers.startTime[markerIndexB])
        )
      );
    });
    expect(container.querySelector('.timelineTrackScreenshotImg')).toBeFalsy();
  });

  it('is created in the <Timeline /> with a profile with screenshots', function () {
    setup(getScreenshotTrackProfile(), <Timeline />);

    // The function `getAllByText` throws already if none are found, with a useful Error,
    // if it can't find any elements. But we still use `expect` to keep a "test-like"
    // assertion, even if it's useless.
    expect(screen.getAllByRole('button', { name: 'Screenshots' })).toHaveLength(
      3
    );

    const screenshotTracks = document.querySelectorAll(
      '.timelineTrackScreenshot'
    );
    expect(screenshotTracks).toHaveLength(3);

    // Tracks 1 and 3 are rendered in the full length because these windows are
    // not closed. They should have the same number of elements.
    expect(screenshotTracks[0].childElementCount).toBe(
      screenshotTracks[2].childElementCount
    );

    // Track 2 is closed before the end and therefore should have less DOM elements.
    expect(screenshotTracks[1].childElementCount).toBeLessThan(
      screenshotTracks[0].childElementCount
    );
  });

  it('is not created in the <Timeline /> with a profile with no screenshots', function () {
    const { queryByText } = setup(getProfileWithNiceTracks(), <Timeline />);
    expect(queryByText('Screenshots')).not.toBeInTheDocument();
  });
});

function setup(
  profile: Profile = getScreenshotTrackProfile(),
  component = <TimelineTrackScreenshots threadIndex={0} windowId="0" />
) {
  const store = storeWithProfile(profile);
  const { getState, dispatch } = store;
  const flushRafCalls = mockRaf();

  // getBoundingClientRect is already mocked by autoMockElementSize.
  jest.spyOn(HTMLElement.prototype, 'getClientRects').mockImplementation(() => {
    return [
      new DOMRect(
        LEFT,
        TOP,
        LEFT + TRACK_WIDTH,
        TOP + FULL_TRACK_SCREENSHOT_HEIGHT
      ),
    ];
  });

  const renderResult = render(<Provider store={store}>{component}</Provider>);
  const { container } = renderResult;

  // WithSize uses requestAnimationFrame
  flushRafCalls();

  function screenshotHover() {
    return ensureExists(
      document.querySelector('.timelineTrackScreenshotHover'),
      `Couldn't find the screenshot hover element, with selector .timelineTrackScreenshotHover`
    );
  }

  function screenshotHoverImage() {
    return ensureExists(
      document.querySelector('.timelineTrackScreenshotHoverImg'),
      `Couldn't find the screenshot hover element, with selector .timelineTrackScreenshotHoverImg`
    );
  }

  function selectionOverlay() {
    return ensureExists(
      document.querySelector('.timelineSelectionOverlay'),
      `Couldn't find the selection element, with selector .timelineSelectionOverlay`
    );
  }

  function screenshotClick(pageX: number) {
    fireFullClick(screenshotTrack(), { pageX, pageY: TOP });
  }

  function screenshotTrack() {
    return ensureExists(
      container.querySelector('.timelineTrackScreenshot'),
      `Couldn't find the screenshot track, with selector .timelineTrackScreenshot`
    );
  }

  function moveMouse(pageX: number) {
    fireEvent(
      screenshotTrack(),
      getMouseEvent('mousemove', { pageX, pageY: TOP, buttons: 1 })
    );
  }

  function moveMouseAndGetLeft(pageX: number): number {
    moveMouse(pageX);
    return parseInt(screenshotHover().style.left);
  }

  function moveMouseAndGetTop(pageX: number): number {
    moveMouse(pageX);
    return parseInt(screenshotHover().style.top);
  }

  function moveMouseAndGetImageSize(pageX: number): {|
    width: number,
    height: number,
  |} {
    moveMouse(pageX);
    const style = screenshotHoverImage().style;
    return {
      width: parseInt(style.width),
      height: parseInt(style.height),
    };
  }

  return {
    ...renderResult,
    dispatch,
    getState,
    thread: profile.threads[0],
    store,
    screenshotHover,
    screenshotTrack,
    moveMouse,
    moveMouseAndGetLeft,
    moveMouseAndGetTop,
    moveMouseAndGetImageSize,
    selectionOverlay,
    screenshotClick,
  };
}

/**
 * Take a thread full screenshot markers, and set some to "Unknown" in order to
 * create gaps in a screenshot track.
 */
function _setScreenshotMarkersToUnknown(
  thread: RawThread,
  shared: RawProfileSharedData,
  ...markerIndexes: IndexIntoRawMarkerTable[]
) {
  // Remove off the last few screenshot markers
  const stringTable = StringTable.withBackingArray(shared.stringArray);
  const unknownStringIndex = stringTable.indexForString('Unknown');
  const screenshotStringIndex = stringTable.indexForString(
    'CompositorScreenshot'
  );
  for (const markerIndex of markerIndexes) {
    // Double check that we've actually got screenshot markers:
    if (thread.markers.name[markerIndex] !== screenshotStringIndex) {
      throw new Error('This is not a screenshot marker.');
    }
    thread.markers.name[markerIndex] = unknownStringIndex;
    thread.markers.data[markerIndex] = null;
  }
}
