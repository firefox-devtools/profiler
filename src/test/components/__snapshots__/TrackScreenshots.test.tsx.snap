// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`timeline/TrackScreenshots matches the component snapshot 1`] = `
<div
  class="timelineTrackScreenshot"
  style="height: 50px;"
>
  <div
    class="timelineTrackScreenshotImgContainer"
    style="left: 0px; width: 37.5px;"
  >
    <img
      class="timelineTrackScreenshotImg"
      src="CompositorScreenshot"
      style="width: 100px; height: 50px;"
    />
  </div>
  <div
    class="timelineTrackScreenshotImgContainer"
    style="left: 37.5px; width: 37.5px;"
  >
    <img
      class="timelineTrackScreenshotImg"
      src="CompositorScreenshot"
      style="width: 100px; height: 50px;"
    />
  </div>
  <div
    class="timelineTrackScreenshotImgContainer"
    style="left: 75px; width: 37.5px;"
  >
    <img
      class="timelineTrackScreenshotImg"
      src="CompositorScreenshot"
      style="width: 100px; height: 50px;"
    />
  </div>
  <div
    class="timelineTrackScreenshotImgContainer"
    style="left: 112.5px; width: 37.5px;"
  >
    <img
      class="timelineTrackScreenshotImg"
      src="CompositorScreenshot"
      style="width: 100px; height: 50px;"
    />
  </div>
  <div
    class="timelineTrackScreenshotImgContainer"
    style="left: 150px; width: 37.5px;"
  >
    <img
      class="timelineTrackScreenshotImg"
      src="CompositorScreenshot"
      style="width: 100px; height: 50px;"
    />
  </div>
  <div
    class="timelineTrackScreenshotImgContainer"
    style="left: 187.5px; width: 37.5px;"
  >
    <img
      class="timelineTrackScreenshotImg"
      src="CompositorScreenshot"
      style="width: 100px; height: 50px;"
    />
  </div>
  <div
    class="timelineTrackScreenshotImgContainer"
    style="left: 225px; width: 37.5px;"
  >
    <img
      class="timelineTrackScreenshotImg"
      src="CompositorScreenshot"
      style="width: 100px; height: 50px;"
    />
  </div>
  <div
    class="timelineTrackScreenshotImgContainer"
    style="left: 262.5px; width: 37.5px;"
  >
    <img
      class="timelineTrackScreenshotImg"
      src="CompositorScreenshot"
      style="width: 100px; height: 50px;"
    />
  </div>
  <div
    class="timelineTrackScreenshotImgContainer"
    style="left: 300px; width: 37.5px;"
  >
    <img
      class="timelineTrackScreenshotImg"
      src="CompositorScreenshot"
      style="width: 100px; height: 50px;"
    />
  </div>
  <div
    class="timelineTrackScreenshotImgContainer"
    style="left: 337.5px; width: 37.5px;"
  >
    <img
      class="timelineTrackScreenshotImg"
      src="CompositorScreenshot"
      style="width: 100px; height: 50px;"
    />
  </div>
  <div
    class="timelineTrackScreenshotImgContainer"
    style="left: 375px; width: 25px;"
  >
    <img
      class="timelineTrackScreenshotImg"
      src="CompositorScreenshot"
      style="width: 100px; height: 50px;"
    />
  </div>
</div>
`;
